src/checkengine/http/mod.rs

```diff
@@ -0,0 +1,220 @@
+use std::io;
+use std::net::{IpAddr, ToSocketAddrs};
+use std::pin::Pin;
+use std::time::{Duration, Instant};
+
+use futures::compat::Future01CompatExt;
+use futures::future::{self, join_all, Future, FutureExt};
+
+use hyper::header::HeaderName;
+use hyper::Uri;
+use hyper_tls::HttpsConnector;
+use native_tls::{self, TlsConnector};
+
+use super::{Check, CheckResult, CheckResultElement, CheckStatus, CheckTimeSpan};
+
+mod hyper_helpers;
+use self::hyper_helpers::{HttpsConnectorWrapped, StaticResolverSingle};
+
+const CHECK_NAME: &str = "http";
+
+#[derive(Clone, Debug)]
+pub struct HttpTarget {
+    pub url: Uri,
+    pub extra_headers: Vec<(HeaderName, String)>,
+}
+
+#[derive(Clone, Debug)]
+pub struct HttpCheck {
+    pub latency_requirement: Duration,
+    pub allow_client_error: bool,
+}
+
+fn invalid_target_missing_value(
+    url: &str,
+    field: &str,
+) -> Pin<Box<dyn Future<Output = Result<CheckResult, io::Error>> + Send + 'static>> {
+    let msg = format!("invalid target: URL {} is missing a {}", url, field);
+    debug!("invalid_target_missing_value: {}", msg);
+    future::err(io::Error::new(io::ErrorKind::Other, msg)).boxed()
+}
+
+fn invalid_target_invalid_value(
+    url: &str,
+    field: &str,
+    value: &str,
+) -> Pin<Box<dyn Future<Output = Result<CheckResult, io::Error>> + Send + 'static>> {
+    let msg = format!(
+        "invalid target: URL {} has invalid {}: {}",
+        url, field, value
+    );
+    debug!("invalid_target_invalid_value: {}", msg);
+    future::err(io::Error::new(io::ErrorKind::Other, msg)).boxed()
+}
+
+impl Check for HttpCheck {
+    type Target = HttpTarget;
+    type Future = Pin<Box<dyn Future<Output = Result<CheckResult, io::Error>> + Send>>;
+
+    fn check(&self, target: Self::Target) -> Self::Future {
+        use http::uri::Parts;
+        let uri_string = target.url.to_string();
+
+        let parts: Parts = target.url.clone().into();
+        let scheme = match parts.scheme {
+            Some(s) => s,
+            None => return invalid_target_missing_value(&uri_string, "scheme"),
+        };
+        let authority = match parts.authority {
+            Some(a) => a,
+            None => return invalid_target_missing_value(&uri_string, "authority"),
+        };
+
+        let mut port;
+        match scheme.as_str() {
+            "http" => port = 80,
```

Comment:
Should these be configurable?

Reply:
no, they are just defaults - if you pass the port in the URL, the port in the URL is used.  But these actually don't go anywhere anyway.  Hyper's resolver wants port numbers but then doesn't use them for requests since it duplicates this logic internally.